pkg/kubelet/cm/container_manager_linux.go

```diff
@@ -342,6 +345,137 @@ func setupKernelTunables(option KernelTunableBehavior) error {
+			// limits are not set for Best Effort pods
+			continue
+		}
+		for _, container := range pod.Spec.Containers {
+			podMemoryLimit += container.Resources.Limits.Memory().Value()
+		}
+		qosMemoryLimits[qosClass] += podMemoryLimit
+	}
+
+	allocatable := cm.nodeInfo.Status.Allocatable.Memory().Value()
+	if allocatable == 0 {
+		return fmt.Errorf("memory capacity reported as 0, might be in standalone mode")
+	}
+
+	for qos, limits := range qosMemoryLimits {
+		glog.V(2).Infof("[Container Manager] %s pod limits total %d bytes (reserve %d%%)", qos, limits, reservePercent)
+	}
+
+	// Calculate QOS memory limits
+	burstableLimit := allocatable - (qosMemoryLimits[v1.PodQOSGuaranteed] * reservePercent / 100)
+	if burstableLimit < minimumLimit {
+		burstableLimit = minimumLimit
+	}
+	bestEffortLimit := burstableLimit - (qosMemoryLimits[v1.PodQOSBurstable] * reservePercent / 100)
+	if bestEffortLimit < minimumLimit {
+		bestEffortLimit = minimumLimit
+	}
+
+	// Update QOS class cgroups with calculated memory limits
+	subsystems, err := GetCgroupSubsystems()
+	if err != nil {
+		return err
+	}
+
+	cgroupManager := NewCgroupManager(subsystems, cm.CgroupDriver)
+	err = setQOSCgroupMemoryLimit(cgroupManager, v1.PodQOSBurstable, burstableLimit)
+	if err != nil {
+		return err
+	}
+	err = setQOSCgroupMemoryLimit(cgroupManager, v1.PodQOSBestEffort, bestEffortLimit)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type qosReserveLimitHandler func(*containerManagerImpl, int64) error
+
+var resourceQOSReserveLimitHandlers = map[string]qosReserveLimitHandler{
+	"memory": setQOSMemoryReserveLimit,
+}
+
+// If usage is already above the limit, set the limit to the usage plus this margin
+// This increases the likelyhood that the limit can be set successfully and gives the
+// cgroup a little forgiveness so it doesn't OOM kill in response to a little growth.
+const memoryUsageMargin = 10 * 1024 * 1024 // 10MB
+
+// The minumum limit to which a QoS level cgroup can be set
+const minimumLimit = 512 * 1024 * 1024 // 512MB
+
+func (cm *containerManagerImpl) setQOSReserveLimits() error {
+	for _, reserveLimit := range cm.NodeConfig.QOSReserveLimits {
+		handler, ok := resourceQOSReserveLimitHandlers[reserveLimit.resource]
+		if !ok {
+			// programming error if we hit this
+			return fmt.Errorf("no reserve limit hanlder for resource '%s'", reserveLimit.resource)
+		}
+		if err := handler(cm, reserveLimit.reservePercent); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Admit always returns Admit: true but allows the ConatinerManager to hook into the admit path
+func (cm *containerManagerImpl) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitResult {
+	glog.V(2).Infof("[Container Manager] Updating QoS reserve limits on pod admission")
+	if cm.CgroupsPerQOS {
+		if err := cm.setQOSReserveLimits(); err != nil {
+			glog.Warningf("[ContainerManager] Failed to set QoS reserve limits at pod admission: %v", err)
```

Comment:
it is a warning so that hopefully conveys that the kubelet can proceed without adverse effects.  not sure how to change the words though without convoluting the meaning.